avogadro/tdxcontroller.h

```diff
@@ -0,0 +1,162 @@
+/******************************************************************************
+  This source file is part of the Avogadro project.
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#ifdef _3DCONNEXION
+#ifndef AVOGADRO_TDXCONTROLLER_H 
+#define AVOGADRO_TDXCONTROLLER_H
+
+#include <SpaceMouse/CImage.hpp>
+#include <SpaceMouse/CNavigation3D.hpp>
+
+#include <QtGui/QImage>
+#include <QtCore/QList>
+#include <QtCore/QMap>
+#include <QtCore/QString>
+#include <QtCore/QList>
+#include "mainwindow.h"
+#include <array>
+
+constexpr uint32_t rayCount = 50;
+
+class QAction;
+
+namespace Avogadro {
+
+namespace Rendering {
+class Camera;
+class GLRenderer;
+}
+
+namespace QtOpenGL {
+class GLWidget;
+}
+
+namespace QtGui {
+class Molecule;
+class ToolPlugin;
+}
+
+
+/**
+ * This class is responsible for handling the TDx navigation in Avogadro2.
+ */
+class TDxController : private TDx::SpaceMouse::Navigation3D::CNavigation3D, private QObject
+{
+
+
+public:
+  TDxController(MainWindow* const mainWindow,
+                Avogadro::QtOpenGL::GLWidget* const pGLWidget,
+                QtGui::Molecule** ppMolecule);
```

Comment:
I don't understand the double pointer. Is there a good reason for this?

Reply:
The double pointer is used to keep track of the actual pointer to active molecule. Then there is no need to update the pointer within the navigation object  each time the active molecule is switched in the explorer. Is there a different approach I should folllow?